src/routes/webhook/webhook-receiver-post.ts

```diff
@@ -0,0 +1,77 @@
+import { BinaryLike, createHmac } from "crypto";
+import { Request, Response } from "express";
+import { validationResult } from "express-validator";
+import { getLogger } from "~/src/config/logger";
+import { GitHubServerApp } from "~/src/models/git-hub-server-app";
+import { WebhookContext } from "./webhook-context";
+
+export const WebhookReceiverPost = async (request: Request, response: Response): Promise<void> => {
+	const errors = validationResult(request);
+	if (!errors.isEmpty()) {
+		response.status(400).json({ errors: errors.array() });
+		return;
+	}
+	const logger = getLogger("webhook.receiver");
+	const eventName = request.headers["x-github-event"] as string;
+	const signatureSHA256 = request.headers["x-hub-signature-256"] as string;
+	const id = request.headers["x-github-delivery"] as string;
+	const uuid = request.params.uuid;
+	let webhookSecret: string = process.env.WEBHOOK_SECRET!;
+	const payload = request.body;
+	try {
+		if (uuid != "cloud") {
+			const gitHubServerApp = await GitHubServerApp.findForUuid(uuid);
+			if (!gitHubServerApp) {
+				response.status(400).send("GitHub app not found");
+				return;
+			}
+			webhookSecret = gitHubServerApp?.webhookSecret;
+		}
+		const verification = createHash(JSON.stringify(payload), webhookSecret);
+		if (verification != signatureSHA256) {
+			response.status(400).send("signature does not match event payload and secret");
+			return;
+		}
+
+		const webhookContext = new WebhookContext({
+			id: id,
+			name: eventName,
+			payload: payload,
+			log: logger
+		});
+
+		const action = "action" in payload ? payload.action : null;
+		if (action) {
+			invokeEventHandler(`${eventName}.${action}`, webhookContext);
+		}
+		invokeEventHandler(eventName, webhookContext);
+
+		response.sendStatus(204);
+
+	} catch (error) {
+		response.sendStatus(500);
+		logger.error(error);
+	}
+};
+
+const invokeEventHandler = (event: string, context: WebhookContext) => {
```

Comment:
I'd probably call this `webhookRouter` or something and would pass in `context:WebhookContext, event:string, action?:string` so that the logic to do both top level event and event.action is in this router instead of the request handler.

Reply:
actually, we can just pass in `WebhookContext` only if we add the action in there as well and let the router decipher who to call.